### PR TITLE
Hack in SimpleNamespace-like functionality to Options

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

This is an awful hack to support various SimpleNamespace-like operations
that qiskit-experiments are using.  We need to transition downstream
packages to the new Options API, but in the meantime, this makes the
Options class behave as much like a SimpleNamespace as it can, while
still keeping the new instance attributes and methods.

### Details and comments

This passes the `qiskit-experiments` tests on my machine.  The real problem is how `qiskit-experiments` accesses `options.__dict__`, so this involves some terrible behaviour to make that work as expected.

It's an absolutely horrific hack, and I regret ever learning Python.

@eggerdj - this passes the tests on my machine, but could you check it does everything you need as well?

### CODE REVIEW

**Review by mtreinish — DISMISSED** [review]
This LGTM, thanks for figuring all these hairy things out. Before we tag it `automerge` and put it in the merge queue though we should verify it works for qiskit-experiments. @eggerdj @nkanazawa1989 @chriseclectic if you could test this that would be great.

**Review by chriseclectic — APPROVED** [review]
I ran qiskit-experiments tests locally with this branch and had no failures

File: test/python/providers/test_options.py [hunk 1]
```diff
@@ -105,3 +108,87 @@ def test_hasattr(self):
...
+    def test_unpacking_dict(self):
+        kwargs = {"hello": "world", "a": "b"}
+        options = Options(**kwargs)
+        self.assertEqual(options.__dict__, kwargs)
+        self.assertEqual({**options.__dict__}, kwargs)
+
+    def test_setting_attributes(self):
+        options = Options()
+        options.hello = "world"
+        options.a = "b"
```
Comment (kdk) [inline test/python/providers/test_options.py]: `test/python/providers/test_options.py:125:8: E0237: Assigning to attribute 'hello' not defined in class slots (assigning-non-slot)` from https://dev.azure.com/qiskit-ci/qiskit-terra/_build/results?buildId=33135&view=logs&j=f0ab2c38-58fa-55ad-8e8b-d3984171f737&t=980625fc-4a64-50e4-7ca0-679c2a4bbda8&l=15

**Comment by jakelishman** [comment]
It's possible that `qiskit-experiments`' CI may fail after this on the linting stage with pylint's `assigning-non-slots`.  If so, you may want to consider just temporarily suppressing that warning; to some degree, it's kind of a pylint bug, because I overrode `__setattr__` to make the assignments work.

File: test/python/providers/test_options.py (hunk 1 above)
Comment (jakelishman) [reply test/python/providers/test_options.py]: Fixed by suppressing it in bdcdf64.  The test succeeds, and pylint could realise that this is possible!

**Review by mtreinish — APPROVED** [review]

**Pull request merged**